gst/tensor_transform/tensor_transform.c

```diff
@@ -280,6 +294,87 @@ gst_tensor_transform_get_mode (const gchar * str)
+      case _NNS_INT16: orc_typecast_to (i, o, n, s16, otype); break; \
+      case _NNS_UINT16: orc_typecast_to (i, o, n, u16, otype); break; \
+      case _NNS_INT8: orc_typecast_to (i, o, n, s8, otype); break; \
+      case _NNS_UINT8: orc_typecast_to (i, o, n, u8, otype); break; \
+      case _NNS_FLOAT64: orc_typecast_to (i, o, n, f64, otype); break; \
+      case _NNS_FLOAT32: orc_typecast_to (i, o, n, f32, otype); break; \
+      default: GST_ERROR_OBJECT (filter, "Unsupported type %d", itype); g_assert (0); break; \
+    } \
+  } while (0)
+
+#define orc_operator_func(i,n,v,opfunc) do { \
+    switch ((v)->type) { \
+      case _NNS_INT32: opfunc (s32) ((gpointer) i, (v)->data._int32_t, n); break; \
+      case _NNS_UINT32: opfunc (u32) ((gpointer) i, (v)->data._uint32_t, n); break; \
+      case _NNS_INT16: opfunc (s16) ((gpointer) i, (v)->data._int16_t, n); break; \
+      case _NNS_UINT16: opfunc (u16) ((gpointer) i, (v)->data._uint16_t, n); break; \
+      case _NNS_INT8: opfunc (s8) ((gpointer) i, (v)->data._int8_t, n); break; \
+      case _NNS_UINT8: opfunc (u8) ((gpointer) i, (v)->data._uint8_t, n); break; \
+      case _NNS_FLOAT64: opfunc (f64) ((gpointer) i, (v)->data._double, n); break; \
+      case _NNS_FLOAT32: opfunc (f32) ((gpointer) i, (v)->data._float, n); break; \
+      default: GST_ERROR_OBJECT (filter, "Unsupported type %d", (v)->type); g_assert (0); break; \
+    } \
+  } while (0)
+
+#define orc_operator_div_loop(i,n,val,typename) do { \
+    gsize idx; \
+    typename *data_array = (typename *) (i); \
+    for (idx = 0; idx < (n); ++idx) { \
+      data_array[idx] /= (val); \
+    } \
+  } while (0)
+
+#define orc_operator(i,n,v,op) do { \
+    switch (op) { \
+      case GTT_OP_ADD: orc_operator_func (i, n, v, orc_func_add); break; \
+      case GTT_OP_MUL: orc_operator_func (i, n, v, orc_func_mul); break; \
+      case GTT_OP_DIV: \
+        switch ((v)->type) { \
+          case _NNS_INT32: orc_operator_div_loop (i, n, (v)->data._int32_t, int32_t); break; \
+          case _NNS_UINT32: orc_operator_div_loop (i, n, (v)->data._uint32_t, uint32_t); break; \
+          case _NNS_INT16: orc_operator_div_loop (i, n, (v)->data._int16_t, int16_t); break; \
+          case _NNS_UINT16: orc_operator_div_loop (i, n, (v)->data._uint16_t, uint16_t); break; \
+          case _NNS_INT8: orc_operator_div_loop (i, n, (v)->data._int8_t, int8_t); break; \
+          case _NNS_UINT8: orc_operator_div_loop (i, n, (v)->data._uint8_t, uint8_t); break; \
+          case _NNS_FLOAT64: orc_func_div (f64) ((gpointer) i, (v)->data._double, n); break; \
+          case _NNS_FLOAT32: orc_func_div (f32) ((gpointer) i, (v)->data._float, n); break; \
+          default: GST_ERROR_OBJECT (filter, "Unsupported type %d", (v)->type); g_assert (0); break; \
+        } \
+        break; \
+      default: GST_ERROR_OBJECT (filter, "Unknown operator %d", op); break; \
+    } \
+  } while (0)
+#endif /* HAVE_ORC */
```

Comment:
Have you checked if the following two cases have the same performance with gcc default optimization?
Case 1:
```
for
    switch
        case1
        case2
```
Case 2
```
switch
    case1
        for
    case2
        for
```

Reply:
interesting result,

```
  const guint array_size = 80000;
  const guint loop = 400;

  /* case 1 */
  for (t = 0; t < loop; ++t) {
    type = (t % 4);

    for (i = 0; i < array_size; ++i) {
      switch (type) {
        case 0:
          data_s8[i] /= 2;
          break;
        case 1:
          data_s32[i] /= 2;
          break;
        default:
          break;
      }
    }
  }

  /* case 2 */
  for (t = 0; t < loop; ++t) {
    type = (t % 4);

    switch (type) {
      case 0:
        for (i = 0; i < array_size; ++i) {
          data_s8[i] /= 2;
        }
        break;
      case 1:
        for (i = 0; i < array_size; ++i) {
          data_s32[i] /= 2;
        }
        break;
      default:
        break;
    }
  }
```

- Result (runs each case twice)
```
** Message: div case 2 : 29786
** Message: div case 1 : 48736
** Message: div case 1 : 44946
** Message: div case 2 : 30144
```